### PR TITLE
Api Will Pool

### LINKED ISSUE #290: Add New Tab Will Pool
<b> Requirement: </b>
- Tab Pool, thêm 1 option nữa là Display cả những người "Sắp pool". Display ra những người:
     + Đang pool
     + Sắp pool (là có đóng góp % với giá trị >0% ở một bill account nào đó có End Date From XXX - XXX là mình được lựa chọn)
- Nếu được thì thêm 1 cột là: Contribute - Ghi ra kiểu: Contribute 50% vào Bill account dự án X; 50% bill account dự án Y
image

### PR DESCRIPTION
**Pull request by khaletuanncc**
## Release Notes

- **New Features**
  - Introduced a new asynchronous endpoint to retrieve resources that will be pooled, allowing users to filter results by various parameters such as username, branch IDs, and contribution date ranges.
  - Enhanced data transfer objects (DTOs) to include additional properties for resource notes, project names, and user identifiers, improving the structure of data related to resource management.
  - Added optional branch ID association for user information to provide additional context.

- **Bug Fixes**
  - None

- **Documentation**
  - None

- **Refactor**
  - None

- **Style**
  - None

- **Tests**
  - None

- **Chores**
  - None

- **Revert**
  - None

### CODE REVIEW

File: aspnet-core/src/ProjectManagement.Core/Services/ResourceService/Dto/InputGetAllWillPoolResourceDto.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace ProjectManagement.Services.ResourceService.Dto
+{
+    public class InputGetAllWillPoolResourceDto : GridParam
+    {
+        public string UserName { get; set; }
+        public List<long> BranchIds { get; set; }
+        public List<UserType> UserTypes { get; set; }
+        public DateTime EndChargeDateFrom { get; set; } = DateTime.Now;
+        public DateTime EndChargeDateTo { get; set; } = DateTime.Now.AddMonths(1);
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Core/Services/ResourceService/Dto/InputGetAllWillPoolResourceDto.cs]: ok

File: aspnet-core/src/ProjectManagement.Application/APIs/Resource/ResourceAppService.cs [hunk 2]
```diff
@@ -283,5 +283,11 @@ public async Task DeleteProjectNote(long projectUserId)
         {
             await _resourceManager.DeleteProjectNote(projectUserId);
         }
+        
+        [HttpPost]
+        public async Task<GridResult<GetAllWillPoolResourceDto>> GetAllWillPoolResource(InputGetAllWillPoolResourceDto input)
+        {
+            return await _resourceManager.GetAllWillPoolResource(input);
+        }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Application/APIs/Resource/ResourceAppService.cs:291]: unnecessary

File: aspnet-core/src/ProjectManagement.Core/Services/ResourceService/ResourceManager.cs [hunk 3]
```diff
@@ -1397,5 +1398,80 @@ public async Task<List<UserShortInfoDto>> GetListUserShortInfo(bool onlyActive)
                 .ToList();
             return query;
         }
+
+        public async Task<GridResult<GetAllWillPoolResourceDto>> GetAllWillPoolResource(InputGetAllWillPoolResourceDto input)
+        {
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Core/Services/ResourceService/ResourceManager.cs:1403]: ok

File: aspnet-core/src/ProjectManagement.Core/Services/ResourceService/ResourceManager.cs [hunk 4]
```diff
@@ -1397,5 +1398,80 @@ public async Task<List<UserShortInfoDto>> GetListUserShortInfo(bool onlyActive)
...
+                    Accounts = new AccountDto()
+                    {
+                        ChargeName = lr.ProjectUserBill.AccountName ?? lr.ProjectUserBill.User.FullName,
+                        HeadCount = lr.ProjectUserBill.HeadCount,
+                        EndChargeDate = lr.ProjectUserBill.EndTime,
+                        Contribute = lr.Contribute,
+                        Note = lr.ProjectUserBill.Note
+                    }
+                })
+                .ToListAsync();
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Core/Services/ResourceService/ResourceManager.cs]: ok

File: aspnet-core/src/ProjectManagement.Core/Services/ResourceService/ResourceManager.cs [hunk 5]
```diff
@@ -1397,5 +1398,80 @@ public async Task<List<UserShortInfoDto>> GetListUserShortInfo(bool onlyActive)
...
+        {
+            // handle filter will pool
+            var query = _workScope.GetAll<LinkedResource>()
+                .Where(lr => lr.Contribute > 0)
+                .Where(lr => lr.ProjectUserBill.EndTime >= input.EndChargeDateFrom && lr.ProjectUserBill.EndTime <= input.EndChargeDateTo)
+                .WhereIf(input.UserName.HasValue(), lr => lr.User.UserName.Contains(input.UserName))
+                .WhereIf(input.BranchIds != null && input.BranchIds.Any(),
+                    lr => lr.User.BranchId.HasValue && input.BranchIds.Contains(lr.User.BranchId.Value))
+                .WhereIf(input.UserTypes != null && input.UserTypes.Any(),
+                    lr => input.UserTypes.Contains(lr.User.UserType));
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Core/Services/ResourceService/ResourceManager.cs]: validation in function ValidationInputFilters

File: aspnet-core/src/ProjectManagement.Core/Services/ResourceService/Dto/GetAllWillPoolResourceDto.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using ProjectManagement.Services.ProjectUserBill.Dto;
+
+namespace ProjectManagement.Services.ResourceService.Dto
+{
```
Comment (tiennguyenhuuncc) [inline aspnet-core/src/ProjectManagement.Core/Services/ResourceService/Dto/GetAllWillPoolResourceDto.cs:9]:
```
public class GetAllWillPoolResourceDto
    {
        public GetUserInfo Resource { get; set; }
        public string ResourceNote {get; set;}
        public List<AccountDto> Accounts { get; set; }
        public string projectName {get; set;}
        public byte TotalContribute => Accounts.Sum(s => s.Contribute);
    }

    public class AccountDto
    {
        public long Id { get; set; }
        public string projectName {get; set;}
        public string ChargeName { get; set; }
        public float HeadCount { get; set; }
        public DateTime? EndChargeDate { get; set; }
        public byte Contribute { get; set; }
        
    }
```

File: aspnet-core/src/ProjectManagement.Core/Services/ResourceService/ResourceManager.cs [hunk 7]
```diff
@@ -1397,5 +1398,107 @@ public async Task<List<UserShortInfoDto>> GetListUserShortInfo(bool onlyActive)
...
+                {
+                    group.Accounts = group.Accounts
+                        .OrderBy(ac => ac.EndChargeDate)
+                        .ToList();
+                }
+            }
+            return new GridResult<GetAllWillPoolResourceDto>(groupResource, groupResource.Count);
+        }
+        
+        private static void ValidationInputFilters(InputGetAllWillPoolResourceDto input)
```
Comment (tiennguyenhuuncc) [inline aspnet-core/src/ProjectManagement.Core/Services/ResourceService/ResourceManager.cs]: default value này thì xly ở phía FE thôi
BE valid endChargeDatTo phải >= endChargeDateFrom là đủ

Comment (tiennguyenhuuncc) [reply aspnet-core/src/ProjectManagement.Core/Services/ResourceService/ResourceManager.cs]: chú ý là FE phải gửi lên dạng yyyy-MM-dd

File: aspnet-core/src/ProjectManagement.Core/Services/ResourceService/ResourceManager.cs [hunk 8]
```diff
@@ -1397,5 +1398,88 @@ public async Task<List<UserShortInfoDto>> GetListUserShortInfo(bool onlyActive)
                 .ToList();
             return query;
         }
+
+        public async Task<GridResult<GetAllWillPoolResourceDto>> GetAllWillPoolResource(InputGetAllWillPoolResourceDto input)
+        {
+            var startDate = input.EndChargeDateFrom.Date;
+            var endDate = input.EndChargeDateTo.Date.AddDays(1).AddTicks(-1);
+            // query get all linked resource
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Core/Services/ResourceService/ResourceManager.cs:1406]: handle in frontend side

**Review by tiennguyenhuuncc — APPROVED** [review]

**Pull request merged**